adc/ade7913/ade7913.c

```diff
@@ -0,0 +1,304 @@
+/*
+ * Phoenix-RTOS
+ *
+ * i.MX RT1176 ADE7913 API
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Marcin Baran
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+
+#include <errno.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <sys/msg.h>
+
+#include <imxrt-multi.h>
+
+#include "ade7913.h"
+
+
+/* ADE7913 registers address */
+#define ADE7913_IWV          0x0
+#define ADE7913_V1WV         0x1
+#define ADE7913_V2WV         0x2
+#define ADE7913_ADC_CRC      0x4
+#define ADE7913_CTRL_CRC     0x5
+#define ADE7913_CNT_SNAPSHOT 0x7
+#define ADE7913_CONFIG       0x8
+#define ADE7913_STATUS0      0x9
+#define ADE7913_LOCK         0xa
+#define ADE7913_SYNC_SNAP    0xb
+#define ADE7913_COUNTER0     0xc
+#define ADE7913_COUNTER1     0xd
+#define ADE7913_EMI_CTRL     0xd
+#define ADE7913_STATUS1      0xf
+#define ADE7913_TEMPOS       0x18
+
+/* ADE7913 registers bits */
+#define ADE7913_CONFIG_CLKOUT_EN (1 << 0)
+#define ADE7913_CONFIG_PWRDWN_EN (1 << 2)
+#define ADE7913_CONFIG_TEMP_EN   (1 << 3)
+#define ADE7913_CONFIG_ADC_FREQ  (1 << 4)
+#define ADE7913_CONFIG_SWRST     (1 << 6)
+#define ADE7913_CONFIG_BW        (1 << 7)
+
+#define ADE7913_STATUS0_RESET_ON (1 << 0)
+#define ADE7913_STATUS0_CRC_STAT (1 << 1)
+#define ADE7913_STATUS0_IC_PROT  (1 << 2)
+
+#define ADE7913_SYNC_SNAP_SYNC (1 << 0)
+#define ADE7913_SYNC_SNAP_SNAP (1 << 1)
+
+#define ADE7913_EMI_CTRLS(X) (1 << X)
+
+#define ADE7913_STATUS1_VERSION (1 << 0)
+#define ADE7913_STATUS1_ADC_NA  (1 << 3)
+
+/* ADE7913 communication utils */
+#define ADE7913_READ_BIT         0x4
+#define ADE7913_ADDR_OFFS        0x3
+#define ADE7913_LOCK_SEQ         0xca
+#define ADE7913_UNLOCK_SEQ       0x9c
+#define ADE7913_CONFIG_FREQ_8KHZ (0x00 << ADE7913_CONFIG_ADC_FREQ)
+#define ADE7913_CONFIG_FREQ_4KHZ (0x01 << ADE7913_CONFIG_ADC_FREQ)
+#define ADE7913_CONFIG_FREQ_2KHZ (0x10 << ADE7913_CONFIG_ADC_FREQ)
+#define ADE7913_CONFIG_FREQ_1KHZ (0x11 << ADE7913_CONFIG_ADC_FREQ)
+#define ADE7913_READY            1
+
+
+static int lpspi_transaction(oid_t *device, int cs,
+	uint8_t *buff, size_t size)
+{
+	msg_t msg;
+	multi_i_t *imsg = (multi_i_t *)msg.i.raw;
+	multi_o_t *omsg = (multi_o_t *)msg.o.raw;
+
+	msg.type = mtDevCtl;
+	msg.i.data = buff;
+	msg.i.size = size;
+	msg.o.data = buff;
+	msg.o.size = size;
+
+	imsg->id = device->id;
+	imsg->spi.type = spi_transaction;
+	imsg->spi.transaction.cs = cs;
+	imsg->spi.transaction.frameSize = size;
+
+	msgSend(device->port, &msg);
+
+	return omsg->err;
+}
+
+
+static int ade7913_read_reg(oid_t *device, int cs,
+	uint8_t addr, uint8_t *data, size_t size)
+{
+	uint8_t buff[size + 1];
```

Comment:
VLA are allowed?

Reply:
VLA is on, not sure it is best approach. @agkaminski , @nalajcie what do you guys think?

Reply:
Based on my knowledge VLA are C99 extension. We try to adhere to ANSI C (C89) standard as much as possible, because of possibility that we'll need to compile our code for some obscure target in the future.

It would be best to allocate this buffer by some other means.